witness/golang/internal/witness/db_test.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"testing"
+
+	_ "github.com/mattn/go-sqlite3" // Load drivers for sqlite3
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// TestRoundTrip tests the happy paths (adding stuff and looking it up).
+func TestRoundTrip(t *testing.T) {
+	for _, test := range []struct {
+		desc      string
+		logID     string
+		c         Chkpt
+		extraRuns int
+	}{
+		{
+			desc:  "simple",
+			logID: "monkeys",
+			c: Chkpt{
+				Size: 123,
+				Raw:  []byte("bananas"),
+			},
+		}, {
+			desc:  "check no failure on inserting same data twice",
+			logID: "monkeys",
+			c: Chkpt{
+				Size: 123,
+				Raw:  []byte("bananas"),
+			},
+			extraRuns: 1,
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+			ctx := context.Background()
+			db, err := sql.Open("sqlite3", ":memory:")
+			if err != nil {
+				t.Error("failed to open temporary in-memory DB", err)
+			}
+			defer db.Close()
+
+			d, err := NewDatabase(db)
+			if err != nil {
+				t.Error("failed to create DB", err)
+			}
+
+			prevChkpts := [2]uint64{0, test.c.Size}
+			for i := 0; i < test.extraRuns+1; i++ {
+				if err := d.SetCheckpoint(ctx, test.logID, prevChkpts[i], &test.c); err != nil {
+					t.Error("failed to set checkpoint", err)
+				}
+				got, err := d.GetLatest(test.logID)
+				if err != nil {
+					t.Error("failed to get latest", err)
+				}
+				if got.Size != test.c.Size || !bytes.Equal(got.Raw, test.c.Raw) {
+					t.Errorf("got != want (%x, %x)", got.Size, test.c.Size)
+				}
+			}
+		})
+	}
+}
+
+// TestUnknownKey should fail because nothing has been added.
+func TestUnknownKey(t *testing.T) {
+	db, err := sql.Open("sqlite3", ":memory:")
+	if err != nil {
+		t.Error("failed to open temporary in-memory DB", err)
+	}
+	defer db.Close()
+
+	d, err := NewDatabase(db)
+	if err != nil {
+		t.Error("failed to create DB", err)
+	}
```

Comment:
This bit is repeated many times in this file. Consider factoring it out to a function, e.g.:

```golang
func mustCreateDB(t *testing.T) (*Database, func() error) {
	t.Helper()
	db, err := sql.Open("sqlite3", ":memory:")
	if err != nil {
		t.Fatalf("failed to open temporary in-memory DB: %v", err)
	}
	d, err := NewDatabase(db)
	if err != nil {
		t.Fatatlf("failed to create DB: %v", err)
	}
	return d, db.Close
}
```

```golang
d, close := mustCreateDB(t)
defer close()
```

Reply:
Oh sorry, `close` is a reserved name, use something different like `closeFn`, or `done`.

Reply:
Done, thank you!